core/src/main/java/org/apache/calcite/rel/rules/IntersectToSemiJoinRule.java

```diff
@@ -0,0 +1,108 @@
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Intersect;
+import org.apache.calcite.rel.core.JoinRelType;
+import org.apache.calcite.rel.logical.LogicalIntersect;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Planner rule that translates a {@link org.apache.calcite.rel.core.Intersect}
+ * to a series of {@link org.apache.calcite.rel.core.Join} that type is
+ * {@link org.apache.calcite.rel.core.JoinRelType#SEMI}.
+ */
+@Value.Enclosing
+public class IntersectToSemiJoinRule
+    extends RelRule<IntersectToSemiJoinRule.Config>
+    implements TransformationRule {
+
+  /** Creates an IntersectToSemiJoinRule. */
+  protected IntersectToSemiJoinRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Intersect intersect = call.rel(0);
+    if (intersect.all) {
+      return; // nothing we can do
+    }
+
+    final RelBuilder builder = call.builder();
+    final RexBuilder rexBuilder = builder.getRexBuilder();
+
+    List<RelNode> inputs = intersect.getInputs();
+    if (inputs.size() != 2) {
+      return;
+    }
+
+    RelNode left = inputs.get(0);
+    RelNode right = inputs.get(1);
+
+    List<RexNode> conditions = new ArrayList<>();
+    int fieldCount = left.getRowType().getFieldCount();
+    for (int i = 0; i < fieldCount; i++) {
+      conditions.add(
+          rexBuilder.makeCall(
+          SqlStdOperatorTable.EQUALS,
+          rexBuilder.makeInputRef(
+              left.getRowType().getFieldList().get(i).getType(), i),
+          rexBuilder.makeInputRef(
+              right.getRowType().getFieldList().get(i).getType(), i + fieldCount)));
+    }
+    RexNode condition = RexUtil.composeConjunction(rexBuilder, conditions);
+
+    RelNode root = builder.push(left)
+        .push(right)
+        .join(JoinRelType.SEMI, condition)
```

Comment:
I am not sure this is enough, you need a distinct somewhere too.
Otherwise this will have duplicates in the outputs if there are duplicates in the inputs.

Reply:
> I am not sure this is enough, you need a distinct somewhere too. Otherwise this will have duplicates in the outputs if there are duplicates in the inputs.

I added distinct after join, which can achieve deduplication. I don't understand your suggestion. Should I add a distinct to each child of join? Can you give me a detailed description?

Reply:
one after or one on each input will work I think

Reply:
> one after or one on each input will work I think

I understand that adding a distinct after join can also ensure deduplication.I still don't understand why I need to add distinct to each child of join.

Reply:
I said "or"

Reply:
> I said "or"

Oh, then my way of writing is also OK, right?

Reply:
Maybe I didn't notice the distinct.

Reply:
> Maybe I didn't notice the distinct.

Maybe yes, thanks. I will add another use case for type inconsistency later. Do you have any other suggestions for me?

Reply:
I would have liked some end-to-end tests (inputs->outputs), but that is not as easy; Quidem tests today cannot invoke optimizations. I keep wanting to work on that subject, it would make everything much easier to test.

Reply:
> I would have liked some end-to-end tests (inputs->outputs), but that is not as easy; Quidem tests today cannot invoke optimizations. I keep wanting to work on that subject, it would make everything much easier to test.

I have studied this test mode and feel it is very good. I want to know what stage Quidem tests has reached now and what I can do for this test mode.

Reply:
in general it's very hard to tell what a plan does; in this case, the plans are simple enough that inspecting them by eye is reasonable.

Reply:
> in general it's very hard to tell what a plan does; in this case, the plans are simple enough that inspecting them by eye is reasonable.

I have an idea. Now that Calcite can convert plans into substrait plans and DuckDB can support the execution of substrait plans, can we complete end-to-end testing? I seem to remember that there is a query engine that also uses DuckDB for result verification.

Reply:
This is a very complicated topic; see this recent thread on the mailing list: https://lists.apache.org/thread/zj0f6fdxdt8v6yjqx5nhsnpjotpo2qtq